packages/contracts/src/RelayHub.sol

```diff
@@ -176,16 +180,20 @@ contract RelayHub is IRelayHub, Ownable {
     {
         RelayCallData memory vars;
         vars.initialGasLeft = aggregateGasleft();
+        vars.relayRequestId = GsnUtils.getRelayRequestID(relayRequest, signature);
         require(!isDeprecated(), "hub deprecated");
         vars.functionSelector = relayRequest.request.data.length>=4 ? MinLibBytes.readBytes4(relayRequest.request.data, 0) : bytes4(0);
-        require(msg.sender == tx.origin, "relay worker must be EOA");
-        vars.relayManager = workerToManager[msg.sender];
-        require(vars.relayManager != address(0), "Unknown relay worker");
-        require(relayRequest.relayData.relayWorker == msg.sender, "Not a right worker");
-        require(
-            isRelayManagerStaked(vars.relayManager),
-            "relay manager not staked"
-        );
+        if (msg.sender != batchGateway){
+            require(signature.length != 0, "missing signature or bad gateway");
```

Comment:
You can, theoretically, use a gateway that is not trusted by this hub and this will cause a 'missing signature' error. Do you prefer it this way?

Reply:
the call comes either from a trusted gateway, or from a relayer.
it means that this signature check should be moved down - report missing signature length only after verfiying it came from a relayer.

Reply:
I don't see a significant difference in where the check is. It will come to the same result. The transaction does not come from a valid Worker => it will be rejected. 
`"relay worker must be EOA"`, `"Not a right worker"`, `"Unknown relay worker"`, `"relay manager not staked"` will all trigger if the transaction is coming from an incorrect BatchGateway.

So I added this `or bad gateway` to the first thrown exception. 
@shahafn what do you think?